system/security/securesocket/securesocket.cpp

```diff
@@ -801,132 +804,105 @@ int CSecureSocket::wait_read(unsigned timeoutms)
 {
-    unsigned elapsed = socketTime(useSeconds) - start;
-    if (elapsed < timeout)
+    size32_t sizeRead = 0;
+    CCycleTimer timer;
+    while (true)
     {
-        unsigned timeleft = timeout - elapsed;
-        if (useSeconds)
-            timeleft *= 1000;
-        return timeleft;
-    }
-    return 0;
-}
-
-void CSecureSocket::readTimeout(void* buf, size32_t min_size, size32_t max_size, size32_t &size_read, unsigned timeout, bool useSeconds)
-{
-    size_read = 0;
-    unsigned start;
-    unsigned timeleft;
-
-    if (timeout != WAIT_FOREVER) {
-        start = socketTime(useSeconds);
-        timeleft = timeout;
-        if (useSeconds)
-            timeleft *= 1000;
-    }
-
-    do {
-        int rc;
-        if (timeout != WAIT_FOREVER) {
-            rc = wait_read(timeleft);
-            if (rc < 0) {
-                THROWSECURESOCKETEXCEPTION(SOCKETERRNO(), "wait_read error"); 
-            }
-            if (rc == 0) {
-                THROWSECURESOCKETEXCEPTION(JSOCKERR_timeout_expired, "timeout expired"); 
-            }
-            timeleft = socketTimeRemaining(useSeconds, start, timeout);
-        }
-
         ERR_clear_error();
-        rc = SSL_read(m_ssl, (char*)buf + size_read, max_size - size_read);
+        int rc = SSL_read(m_ssl, (char*)buf + sizeRead, max_size - sizeRead);
+
         if (rc > 0)
         {
-            size_read += rc;
+            sizeRead += rc;
+            if (sizeRead >= min_size)
+                break;
         }
-        else
+        else if (0 == rc)
         {
-            int err = SSL_get_error(m_ssl, rc);
-            char errbuf[512];
-            ERR_error_string_n(err, errbuf, 512);
-            ERR_clear_error();
-            VStringBuffer errmsg("SSL_read error %d - %s", err, errbuf);
-            if (m_loglevel >= SSLogMax)
-                DBGLOG("Warning: %s", errmsg.str());
-            if (min_size > 0)
-                throw createJSocketException(JSOCKERR_graceful_close, errmsg);
+            if (0 == min_size) // mirror behviour of jsocket impl.
+                break;
+            THROWJSOCKEXCEPTION(JSOCKERR_graceful_close);
         }
-    } while (size_read < min_size);
-}
-
+        else
+        {
+            int ssl_err = SSL_get_error(m_ssl, rc);
+            if (nonBlocking && (ssl_err == SSL_ERROR_WANT_READ || ssl_err == SSL_ERROR_WANT_WRITE)) // NB: SSL_read can cause SSL_ERROR_WANT_WRITE
+            {
+                if (0 == min_size) // if min_read is 0, then whatever we have read so far is good enough (even if 0)
+                    break;
 
-void CSecureSocket::readtms(void* buf, size32_t min_size, size32_t max_size, size32_t &size_read, unsigned timeoutms)
-{
-    readTimeout(buf, min_size, max_size, size_read, timeoutms, false);
+                unsigned remainingMs = timer.remainingMs(timeoutMs);
+                rc = wait_read(remainingMs);
```

Comment:
If we get a SSL_WANT_WRITE error then should we call wait_write() here instead of wait_read() ?

Reply:
I guess so, I haven't seen SSL_WANT_WRITE occur in testing, but.. will change to handle either.